lib/core/src/system/gpu/data/texture.rs

```diff
@@ -326,49 +331,72 @@ with_texture_format_relations!(generate_internal_format_instances []);
 
 
 
-// =====================
-// === TextureSource ===
-// =====================
+// =======================
+// === Texture sources ===
+// =======================
+
+// === Empty Texture ===
+
+/// An empty texture, created as 1x1 texture in Rgba format.
+pub struct EmptyTexture;
+
 
-/// Source of the texture. Please note that the texture will be loaded asynchronously on demand.
-#[derive(Clone,Debug)]
-pub enum TextureSource {
-    /// URL the texture should be loaded from. This source implies asynchronous loading.
-    Url(String)
+
+// === Texture from url ===
+
+/// Texture loaded from given url. It is loaded to gpu asynchronously, putting a mocked texture
+/// before
+pub struct TextureFromUrl {
+    /// An url from where the texture is loaded.
+    pub url       : String,
 }
 
-impl<S:Str> From<S> for TextureSource {
+
+impl<S:Str> From<S> for TextureFromUrl {
     fn from(s:S) -> Self {
-        Self::Url(s.into())
+        Self{url:s.into()}
     }
 }
 
 
+// === Texture from memory ===
+
+/// A texture created from slice.
+pub struct TextureFromMemory<'a,T> {
+    /// A slice from which the texture data are taken.
+    pub view   : &'a[T],
+    /// Texture's width
+    pub width  : i32,
+    /// Texture's height
+    pub height : i32,
+}
+
+
 
 // ===============
 // === Texture ===
 // ===============
 
-/// Texture representation.
-#[derive(Derivative)]
-#[derivative(Clone(bound=""))]
-#[derivative(Debug(bound=""))]
-pub struct Texture<InternalFormat,ElemType> {
-    source  : TextureSource,
-    phantom : PhantomData2<InternalFormat,ElemType>,
+/// A Texture.
+pub struct Texture<Source,InternalFormat,ElemType> {
```

Comment:
I'm not sure if the texture should be parametrized with `Source`. The longer I think about it the more I dislike this idea. Let's have a call about it.

Reply:
Especially if you want to texture data being owned by texture, the parameter should remain. When I want to extend texture data, I need to specify what the source is (extending texture data is valid only for Memory source), or making ugly pattern match with one valid variant and panics elsewhere.